Mempool reference-auditing feature in checked builds

xmcclure opened the pull request:
This introduces three macros: CHECKED_METADATA_WRITE_PTR, CHECKED_METADATA_WRITE_PTR_LOCAL, CHECKED_METADATA_WRITE_PTR_EXEMPT. These can be used in place of a variable assignment [i.e. CHECKED_METADATA_WRITE_PTR(obj->ptr, val) instead of obj->ptr = val], and should be used whenever writing a pointer into memory managed by an image or imageset. In checked mode, the macros verify that the image or imageset properly holds a reference count for the memory being pointed to.

The first commit contains a certain amount of cleanup to support the feature:
- Rework in mempool.c: Have USE_MALLOC_FOR_MEMPOOLS work by tweaking existing rather than using an entire different path. Remove unused mempool field “rest”, stop assigning fields “pos” and “end” where unused. Removed unused function mono_mempool_empty. Fix off-by-one bug (>= vs >) in mono_mempool_contains_addr.
- Fix misc bug spotted by Rodrigo in aot-runtime.c: image of MonoGenericParam was always being set to corlib instead of proper image
- Add comments and replace inline constants with named constants in various places

The second commit adds the audit macro feature itself, and the third commit adds a very small number of  macro invocations.

xmcclure commented on mono/metadata/mempool.c:
```diff
@@ -40,33 +38,42 @@
+
 	union {
-		double pad; /* to assure proper alignment */
+		// Unused: Imposing floating point memory rules on _MonoMemPool's final field ensures proper alignment of whole header struct
```
Just checking-- Is this a correct description of the intent?

xmcclure commented on mono/metadata/mempool.c:
```diff
@@ -137,7 +147,7 @@ mono_mempool_destroy (MonoMemPool *pool)
 mono_mempool_invalidate (MonoMemPool *pool)
 {
-#ifdef MALLOC_ALLOCATION
+#ifdef INDIVIDUAL_ALLOCATIONS
```
Checking-- when I refactored I tried not to change any actual behavior, so I kept in these places where we verify MALLOC_ALLOCATION is not being used in conjunction with these other features. Now that the two paths are unified however I think all of these features should actually work even in INDIVIDUAL_ALLOCATION mode. Could I go ahead and remove the g_assert_not_reached cases like this?

xmcclure commented:
I added a new commit addressing the above issues. I have also a commit that adds more audit points but I am holding this commit back until I make an attempt to fix some of the issues it found >_>

xmcclure commented on mono/metadata/image.c:
```diff
@@ -1171,7 +1205,7 @@ MonoImage *
 {
 	GuidData data;
-	GHashTable *loaded_images = refonly ? loaded_images_refonly_hash : loaded_images_hash;
+	GHashTable *loaded_images = loaded_images_hash (TRUE, refonly);
```
I would welcome any input on whether people think the interface to loaded_images_hash here is a good one.

schani commented on mono/utils/checked-build.h:
```diff
@@ -12,6 +12,10 @@
 #include <config.h>
 
+// This is for metadata writes which do not have to be checked.
+// We still use the macro so that they are easier to find later.
```
An explanation on what the circumstances are under which a write should be exempt would be nice.

xmcclure commented:
Just made these changes to the pull request:
- Rebased on top of the now-merged #2175 
- Dropped the commit with the initial audit points
- Improved the comment pointed out by @schani

kumpera commented on mono/utils/checked-build.c:
```diff
@@ -225,4 +229,279 @@ assert_gc_neutral_mode (void)
+
+			mono_image_lock (checking);
+
+			g_warning ("QQPTX: idx %d this pass, image %p (%s) has %d modules %d references (%p)", (int)current_idx, checking, checking->name, (int)checking->module_count, (int)checking->nreferences, checking->references); // FIXME: Remove all QQPTX lines
```
Don't remove the log messages, you'll miss them when troubleshooting issues.

One pattern we have in the runtime is to hide them under a macro and disable by default.

xmcclure replied:
I messed up the rebase in a couple ways... those logs were NOT meant to be included :/

kumpera commented:
LGTM, merge ready.
